File: learninig_tracker/backend/src/main/java/com/tracker/service/ProjectService.java
package com.tracker.service;

import com.tracker.model.Project;
import com.tracker.repository.ProjectRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class ProjectService {

    private final ProjectRepository projectRepository;

    public List<Project> getAllProjects() {
        return projectRepository.findAll();
    }

    public Optional<Project> getProjectById(UUID id) {
        return projectRepository.findById(id);
    }

    public Project createProject(Project project) {
        return projectRepository.save(project);
    }

    public Optional<Project> updateProject(UUID id, Project projectDetails) {
        return projectRepository.findById(id)
                .map(existingProject -> {
                    existingProject.setName(projectDetails.getName());
                    existingProject.setDescription(projectDetails.getDescription());
                    existingProject.setStatus(projectDetails.getStatus());
                    existingProject.setPriority(projectDetails.getPriority());
                    existingProject.setCategoryId(projectDetails.getCategoryId());
                    existingProject.setEstimatedCompletionDate(projectDetails.getEstimatedCompletionDate());
                    existingProject.setStartDate(projectDetails.getStartDate());
                    existingProject.setCompletionDate(projectDetails.getCompletionDate());
                    return projectRepository.save(existingProject);
                });
    }

    public boolean deleteProject(UUID id) {
        return projectRepository.findById(id)
                .map(project -> {
                    projectRepository.delete(project);
                    return true;
                })
                .orElse(false);
    }

    public List<Project> getProjectsByCategory(UUID categoryId) {
        return projectRepository.findByCategoryId(categoryId);
    }

    public List<Project> getProjectsByStatus(String status) {
        return projectRepository.findByStatus(status);
    }

    public List<Project> getProjectsByPriority(String priority) {
        return projectRepository.findByPriority(priority);
    }

    public List<Project> getActiveProjects() {
        return projectRepository.findByStatusNot("COMPLETED");
    }
}

File: learninig_tracker/frontend/src/services/projectService.ts
import { api } from './api';
import { ProjectDTO, Project } from '../types';

const BASE_URL = '/projects';

export const ProjectService = {
  getAllProjects: async (): Promise<ProjectDTO[]> => {
    const response = await api.get(BASE_URL);
    return response.data;
  },

  getProjectById: async (id: string): Promise<ProjectDTO> => {
    const response = await api.get(`${BASE_URL}/${id}`);
    return response.data;
  },

  createProject: async (project: Omit<Project, 'id'>): Promise<ProjectDTO> => {
    const response = await api.post(BASE_URL, project);
    return response.data;
  },

  updateProject: async (id: string, project: Omit<Project, 'id'>): Promise<ProjectDTO> => {
    const response = await api.put(`${BASE_URL}/${id}`, project);
    return response.data;
  },

  deleteProject: async (id: string): Promise<void> => {
    await api.delete(`${BASE_URL}/${id}`);
  },

  getProjectsByCategory: async (categoryId: string): Promise<ProjectDTO[]> => {
    const response = await api.get(`${BASE_URL}/category/${categoryId}`);
    return response.data;
  },

  getProjectsByStatus: async (status: string): Promise<ProjectDTO[]> => {
    const response = await api.get(`${BASE_URL}/status/${status}`);
    return response.data;
  },

  getProjectsByPriority: async (priority: string): Promise<ProjectDTO[]> => {
    const response = await api.get(`${BASE_URL}/priority/${priority}`);
    return response.data;
  },

  getActiveProjects: async (): Promise<ProjectDTO[]> => {
    const response = await api.get(`${BASE_URL}/active`);
    return response.data;
  },
};
